compiler/noirc_evaluator/src/ssa/ssa_gen/context.rs

```diff
@@ -408,10 +412,19 @@ impl<'a> FunctionContext<'a> {
                 let max_lhs_bits = dfg.get_value_max_num_bits(lhs);
 
                 match operator {
-                    BinaryOpKind::Add | BinaryOpKind::Subtract | BinaryOpKind::Multiply => {
+                    BinaryOpKind::Add | BinaryOpKind::Subtract => {
                         // Overflow check is deferred to acir-gen
                         return result;
                     }
+                    BinaryOpKind::Multiply => {
+                        // Part of the overflow check is done here for u128
+                        if matches!(result_type, NumericType::Unsigned { bit_size: 128 }) {
+                            self.check_u128_mul_overflow(lhs, rhs);
+                        }
+
+                        // Overflow check is also done in acir-gen
+                        return result;
+                    }
```

Comment:
I just noticed that we're doing this up in SSA gen. We want to avoid this happening in brillig functions as the overflow checks in brillig gen should work fine.

Reply:
I think a small SSA pass which just searches for multiplications and then inserts the instructions which are generated by `check_u128_mul_overflow` afterwards would be best. We can run this towards the end of the SSA pipeline which avoids the situation which we have currently where the overflow checks could prevent a multiplication from being optimized out.

Reply:
Eh, I could tolerate this being done in ssa gen if we have an issue to explore moving it further down the SSA pipeline. We should be able to distinguish between runtimes up here at least.